cmake/Modules/CodeCoverage.cmake

```diff
@@ -337,209 +341,347 @@ function(ADD_CODE_COVERAGE)
       COMMAND
-        ${LCOV_PATH} ${LCOV_EXTRA_FLAGS}
-        -a ${PROJECT_BINARY_DIR}/${Coverage_NAME}_base_cpp.info
-        -a ${PROJECT_BINARY_DIR}/${Coverage_NAME}.info
-        --output-file ${PROJECT_BINARY_DIR}/${Coverage_NAME}.info.total
-        || echo "WARNING: No cpp report to output"
+        if [ -s ${PROJECT_BINARY_DIR}/gtest_cpp_coverage/${Coverage_NAME}_gtest_cpp.info.original ]\; then
+          ${LCOV_PATH} ${LCOV_EXTRA_FLAGS}
+          -a ${PROJECT_BINARY_DIR}/${Coverage_NAME}_base_cpp.info
+          -a ${PROJECT_BINARY_DIR}/gtest_cpp_coverage/${Coverage_NAME}_gtest_cpp.info.original
+          --output-file ${PROJECT_BINARY_DIR}/gtest_cpp_coverage/${Coverage_NAME}_gtest_cpp.info.total\;
+        fi
       COMMAND
-        ${LCOV_PATH} ${LCOV_EXTRA_FLAGS}
-        --remove ${PROJECT_BINARY_DIR}/${Coverage_NAME}.info.total ${LCOV_REMOVES}
-        --output-file ${PROJECT_BINARY_DIR}/${Coverage_NAME}.info.removed
-        ||  echo "WARNING: No cpp report to output"
+        if [ -f ${PROJECT_BINARY_DIR}/gtest_cpp_coverage/${Coverage_NAME}_gtest_cpp.info.total ]\; then
+          ${LCOV_PATH} ${LCOV_EXTRA_FLAGS} --remove
+          ${PROJECT_BINARY_DIR}/gtest_cpp_coverage/${Coverage_NAME}_gtest_cpp.info.total
+          ${LCOV_REMOVES}
+          --output-file
+          ${PROJECT_BINARY_DIR}/gtest_cpp_coverage/${Coverage_NAME}_gtest_cpp.info.removed\;
+        fi
       COMMAND
-        ${LCOV_PATH} ${LCOV_EXTRA_FLAGS}
-        --extract ${PROJECT_BINARY_DIR}/${Coverage_NAME}.info.removed "'*${REAL_SOURCE_DIR}*'"
-        --output-file ${PROJECT_BINARY_DIR}/${Coverage_NAME}.info.cleaned
-        || echo "WARNING: No cpp report to output"
+        if [ -f ${PROJECT_BINARY_DIR}/gtest_cpp_coverage/${Coverage_NAME}_gtest_cpp.info.removed ]\; then
+          ${LCOV_PATH} ${LCOV_EXTRA_FLAGS}
+          --extract
+          ${PROJECT_BINARY_DIR}/gtest_cpp_coverage/${Coverage_NAME}_gtest_cpp.info.removed
+          "'*${REAL_SOURCE_DIR}*'"
+          --output-file
+          ${PROJECT_BINARY_DIR}/gtest_cpp_coverage/${Coverage_NAME}_gtest_cpp.info.cleaned\;
+        fi
       COMMAND
-        ${CMAKE_COMMAND} -E make_directory
-        ${PROJECT_BINARY_DIR}/cpp_coverage
-        || echo "WARNING: Error to create cpp coverage dir"
+        if [ -f ${PROJECT_BINARY_DIR}/gtest_cpp_coverage/${Coverage_NAME}_gtest_cpp.info.cleaned ]\; then
+          ${CMAKE_COMMAND} -E copy
+          ${PROJECT_BINARY_DIR}/gtest_cpp_coverage/${Coverage_NAME}_gtest_cpp.info.cleaned
+          ${PROJECT_BINARY_DIR}/${Coverage_NAME}_gtest_cpp.info\;
+        fi
+      WORKING_DIRECTORY ${PROJECT_BINARY_DIR}
+      DEPENDS
+        _run_tests_${PROJECT_NAME}_gtest
+        _create_coverage_report_${PROJECT_NAME}_gtest_cpp
+        ${PROJECT_BINARY_DIR}/gtest_cpp_coverage/${Coverage_NAME}_gtest_cpp.info.original
+        _run_tests_${PROJECT_NAME}_cpp_base_coverage_report
+        ${PROJECT_BINARY_DIR}/${Coverage_NAME}_base_cpp.info
+    )
```

Comment:
Could you separate this `add_custom_command` into a few commands? I think using `if` in multiple `COMMAND` is too complicated.

Reply:
the command is too complicated to me, too.
But I also don't want to split the command and make a lot of custom target.
Now I'm thinking of rewriting the commands into python scripts, and keep the target dependency simple.
@garaemon what do you think?

My plan is.
- move this script to gitai_code_coverage
- covert the add_custom_commands to python script

Reply:
@garaemon 

I splitted this PR into #19 #20 #21 #22 #23.
After #23 is merged, I will work changing this scripts into python.